orderbook/src/api/get_fee_and_price.rs

```diff
@@ -0,0 +1,186 @@
+use super::H160Wrapper;
+use crate::fee::{MinFeeCalculating, MinFeeCalculationError};
+use anyhow::Result;
+use chrono::{DateTime, Utc};
+use ethcontract::{H160, U256};
+use model::{order::OrderKind, u256_decimal};
+use num::BigInt;
+use serde::{Deserialize, Serialize};
+use shared::{
+    conversions::U256Ext,
+    price_estimate::{PriceEstimating, PriceEstimationError},
+};
+use std::convert::Infallible;
+use std::sync::Arc;
+use warp::{hyper::StatusCode, reply, Filter, Rejection, Reply};
+
+#[derive(Deserialize)]
+struct Query {
+    sell_token: H160Wrapper,
+    buy_token: H160Wrapper,
+    #[serde(with = "u256_decimal")]
+    amount: U256,
+    kind: OrderKind,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Response {
+    expiration_date: DateTime<Utc>,
+    min_fee: U256,
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    amount_after_fee: BigInt,
+}
+
+#[derive(Debug)]
+enum Error {
+    NotFound,
+    UnsupportedToken(H160),
+    Other(anyhow::Error),
+}
+
+impl From<MinFeeCalculationError> for Error {
+    fn from(other: MinFeeCalculationError) -> Self {
+        match other {
+            MinFeeCalculationError::NotFound => Error::NotFound,
+            MinFeeCalculationError::UnsupportedToken(token) => Error::UnsupportedToken(token),
+            MinFeeCalculationError::Other(error) => Error::Other(error),
+        }
+    }
+}
+
+impl From<PriceEstimationError> for Error {
+    fn from(other: PriceEstimationError) -> Self {
+        match other {
+            PriceEstimationError::UnsupportedToken(token) => Error::UnsupportedToken(token),
+            PriceEstimationError::Other(error) => Error::Other(error),
+        }
+    }
+}
+
+fn request() -> impl Filter<Extract = (Query,), Error = Rejection> + Clone {
+    warp::path!("fee_and_price")
+        .and(warp::get())
+        .and(warp::query::<Query>())
+}
+
+fn response(result: Result<Response, Error>) -> impl Reply {
+    match result {
+        Ok(response) => reply::with_status(reply::json(&response), StatusCode::OK),
+        Err(Error::NotFound) => reply::with_status(
+            super::error("NotFound", "Token was not found"),
+            StatusCode::NOT_FOUND,
+        ),
+        Err(Error::UnsupportedToken(token)) => reply::with_status(
+            super::error("UnsupportedToken", format!("Token address {:?}", token)),
+            StatusCode::BAD_REQUEST,
+        ),
+        Err(Error::Other(err)) => {
+            tracing::error!(?err, "get_fee_and_price error");
+            reply::with_status(super::internal_error(), StatusCode::INTERNAL_SERVER_ERROR)
+        }
+    }
+}
+
+async fn calculate(
+    fee_calculator: Arc<dyn MinFeeCalculating>,
+    price_estimator: Arc<dyn PriceEstimating>,
+    query: Query,
+) -> Result<Response, Error> {
+    let (min_fee, expiration_date) = fee_calculator
+        .min_fee(query.sell_token.0, None, None, None)
```

Comment:
This should not use the legacy API but instead set buyToken, amount and kind (since it already knows it).

Reply:
This is intentional to make the implementation simpler. With the advanced implementation we need to take into account that the fee for the full sell amount is different from the new fee after deducting the full amount fee. So I want to handle that in a future PR.

Reply:
The frontend currently uses the new API (with the full sell amount which is technically not correct but will still give a better estimate than not providing any information about the route at all). It doesn't have to be accurate as in the "verification" we still use the legacy API (effectively accepting the lowest fee that was provided over all routes/amounts for this sell token). Yet it would be nice if it was as close as possible (the same the frontend uses today)

If left like this, it would regress the amount of fee the orders are providing by default.

Reply:
I didn't realize the verification uses the legacy api. Fixed.